### ISSUE TITLE
Client doesn't switch output devices

### ISSUE THREAD

**Issue by TempleClause** [telegramdesktop/tdesktop#776]
It's not really a bug I guess.
The thing is the sound will always play through the default sound device set when the windows app was started. If you switch the default sound device afterwards it will still play through the default sound device set when the app was started. 
This is annoying because I often switch between speakers and headset.

**Comment by Cyberuben** [comment]
+1

**Comment by Lightdewd** [comment]
I can confirm this issue. Reopening the application fixes this but if it would switch to the default device that would be better!

**Comment by Bullsquid** [comment]
This actually really is a bug, and it's pretty annoying. My sound card has two output devices that switch automatically when I plug in my headphones. So every time I plug or unplug them, telegram sound goes through the wrong device and I have to relaunch it.
The sound should always go through the default device.
P.S. I'm using Windows 10 and TDesktop v0.9.15

**Comment by Atomkraftzwerg** [comment]
+1

**Comment by Bullsquid** [comment]
@Atomkraftzwerg or just use the default output device (surprise) by default.

**Comment by LiamDawe** [comment]
Okay, also an issue here, and damn annoying.

**Comment by matthijs186** [comment]
+1

**Comment by matthijs186** [comment]
Would be happy if Telegram could detect this change.

**Comment by Skaronator** [comment]
Can confirm this issue. Seems like Windows changed something in the audiodriver in 8 or 10. Everything works fine on W7 but not on 10.

**Comment by john-preston** [comment]
@Skaronator I thought it doesn't work anywhere..

**Comment by Bullsquid** [comment]
@john-preston My guess is that @Skaronator has different driver settings in Windows 7 resulting in only one device instead of two.

**Comment by Skaronator** [comment]
I'm on **Windows 10** and I use **SoundSwitch** to switch between my speakers ("Lautsprecher") and headset ("Kopfhörer"). And when I switch between the devices does it still play on the device that was the default device on boot.

![image](https://cloud.githubusercontent.com/assets/2886913/16062446/722027e8-3293-11e6-9e95-90881a87f7b9.png)

**Comment by mirh** [comment]
That's another matter. You shouldn't need soundswitch in the first place for the records. 
And without that, it switching is broken in W7 too.

**Comment by KiDmorbid** [comment]
I can confirm that this issue is still not fixed. I'm also using SoundSwitch and can confirm @Skaronator report but I'm in Windows 7 and not on W10.

**Comment by z3ntu** [comment]
Is it that hard to fix this?

**Comment by Erusy** [comment]
Problem persistent on Windows and Linux, maybe it's in something in tdesktop cross-platform code.

**Comment by z3ntu** [comment]
@Erusy apparently...

**Comment by john-preston** [comment]
@z3ntu This is how OpenAL works, it doesn't offer a way to follow the default audio output device changing :(

**Comment by Erusy** [comment]
Found kind of glitch:
First I configure /etc/openal/alsoft.conf :
`#allow-moves = false` 
to 
`allow-moves = true`
Then restarted tdesktop and launched pavucontrol. 
In tdesktop started recording and while holding LMB moved cursor to pavucontrol window.
While holding LMB pressed RMB — this allowed operate in pavucontrol window and continue recording in tdesktop. Default device was not recognized, but I can pick one another and this worked. 
But for some reason only once, second time sound was recorded (red dot is moving/flashing), but no message formed.

**Comment by z3ntu** [comment]
@Erusy Does changing the output device also work with that method?

**Comment by Erusy** [comment]
On my system tdesktop output stream is always woking, and output device changing also works and this setting survives restart, so I just don't need to use LMB-RMB-cheat.
But I've just tested it and it works too.

**Comment by sjmuniznatgeo** [comment]
Linux+pulseaudio. Telegram shows at the pavucontrol app list, but I am unable to switch its output.
I was willing to listen audio messages on USB headset, but I can't change from normal speakers.
Thanks!

**Comment by Sandvich** [comment]
Also seeing this on Linux + Phonon GStreamer.

**Comment by thesweetiger** [comment]
This is a huge problem with the hdmi audio
When I leave the sleep mode on my laptop sometime my avr is not initialised so windows use the onboard sound card and after return on the hdmi , so everytime i need to relaunch telegram for be able to listen a video or a listen an audio message because telegram is stuck on a non-existent output

I start to be tired of this problem , please fix it 
I have no problem with other software

**Comment by socketpair** [comment]
The same problem

**Comment by MasterGroosha** [comment]
Looks like it's fixed in 1.0.3 alpha!

**Comment by drizzt** [comment]
Not fixed on Linux at all

**Comment by z3ntu** [comment]
For me in 1.0.37, Linux (Arch Linux and Pulseaudio) it's fixed as-in there isn't a constant sound stream active in the "Playback" tab in pavucontrol (equivalent is "Applications" in the Plasma Audio Volume Settings)

**Comment by Kunepro** [comment]
This issue is still there with Telegram for OSX Sierra (all up to date).

**Comment by afliw** [comment]
Any news on this?
It's really annoying to have to relaunch Telegram every time I connect my headphones or I want to record a voice message with Stereo Mix.

**Comment by john-preston** [comment]
@afliw The audio device is recreated each time nothing is playing so it should use default device starting from at least a sound output pause.

**Comment by nimatrueway** [comment]
I have `Telegram 1.1.23 alpha` installed on `Ubuntu 17.10` and can't switch output channel.
```
> pacmd move-sink-input 41 0                                                                                                                                                            
Moved failed.
```

**Comment by stek29** [comment]
See #2231

**Comment by beroal** [comment]
It still does not work on Linux. `pactl move-sink-input $SINK_NUMBER 2` outputs `Failure: Invalid argument` if `SINK_NUMBER` corresponds to "telegram-desktop". If `SINK_NUMBER` corresponds to Audacious, the command works. ArchLinux.

**Comment by fintynator** [comment]
is kinda sad that after almost 4 years now this is not fixed ....are you serious ?

**Comment by senid231** [comment]
@john-preston can we do something with this issue?

at least ability to change preferable input/output device from settings?

**Comment by MasterGroosha** [comment]
@senid231 What OS are you on? TDesktop for Windows already has output options in Settings, can't say for Linux/macOS though

**Comment by mirh** [comment]
On linux if any there's this issue IIRC: https://github.com/telegramdesktop/tdesktop/issues/3459#issuecomment-397711149

**Comment by NMelis** [comment]
I solved this bug by next: the first i disabled (off) all the device except necessary for me
![image](https://user-images.githubusercontent.com/22614042/52815074-44f8d600-30d0-11e9-84be-a90fb434b755.png) to next i can to change device for telegram

**Comment by senid231** [comment]
@MasterGroosha I'm using Linux debian stretch

@mirh it works, but stream is recreated on every play/pause so after resume video/audio you must switch output again

@NMelis I done same thing to make it work

**Comment by saeidscorp** [comment]
I work around this issue by first connecting my bluetooth headphones, restarting the `pulseaudio` daemon with a pause in between, and then reconnecting my bluetooth headphones all over again.

I use this for the restart:
```
pulseaudio -k && sleep 3 && pulseaudio --start
```

After a restart of Telegram, It then works as well and as sane as all other programs.

(Somehow the initial BT connection is necessary before `pulseaudio` restart)

**Comment by facundobatista** [comment]
Maybe the `DONT_MOVE` flag here is related to this? (this flag comes from Telegram when requesting the audio device)

```
$ pacmd list-sink-inputs
2 sink input(s) available.
    index: 25
        (...)
    index: 33
        driver: <protocol-native.c>
        flags: DONT_MOVE START_CORKED
        state: RUNNING
        sink: 0 <alsa_output.usb-Plantronics_Wireless_Audio_Plantronics_Wireless_Audio-00.analog-stereo>
        volume: front-left: 68812 / 105% / 1,27 dB,   front-right: 68812 / 105% / 1,27 dB
                balance 0,00
        muted: no
        current latency: 55,48 ms
        requested latency: 16,00 ms
        sample spec: float32le 2ch 48000Hz
        channel map: front-left,front-right
                     Estéreo
        resample method: speex-float-1
        module: 14
        client: 28 <Telegram>
        properties:
                media.role = hex:
                phonon.streamid = hex:
                media.name = "Playback Stream"
                application.name = "Telegram"
                native-protocol.peer = "UNIX socket client"
                native-protocol.version = "32"
                application.process.id = "2423"
                application.process.user = "facundo"
                application.process.host = "blackfx"
                application.process.binary = "Telegram"
                application.language = "es_AR.UTF-8"
                window.x11.display = ":0"
                application.process.machine_id = "5121f437f343485b960c244df78aa216"
                application.process.session_id = "1"
                module-stream-restore.id = "sink-input-by-application-name:Telegram"
```

**Comment by mirh** [comment]
Inspired by this nice reporting, I followed the sources breadcrumb trail first uselessly in platform/linux/ and then eventually in https://github.com/telegramdesktop/tdesktop/blob/bec39d89e19670eb436dc794a8f20b657cb87c71/Telegram/SourceFiles/media/audio/media_audio.cpp#L133-L136

After checking for pulse audio in OAL, [here you are](https://github.com/kcat/openal-soft/blob/61f7e7716c6743b16051d8c3ea4cea3b27d0197b/Alc/backends/pulseaudio.cpp#L869). 
Which as I already linked is the workaround you get in #3459

**Comment by mirh** [comment]
https://github.com/kcat/openal-soft/commit/c135629eae544b0e7785803b98987c1ab1eb95ca
This should have been fixed whenever you build against openal 1.20.

**Comment by ilya-fedin** [comment]
> This should have been fixed whenever you build against openal 1.20.

This already done 3 months ago
https://github.com/telegramdesktop/tdesktop/issues/2231#event-3113582743
